File: server/storage.ts
import { products, sellIns, sellThroughs, type Product, type InsertProduct, type SellIn, type InsertSellIn, type SellThrough, type InsertSellThrough, type ProductAnalytics, type OverallAnalytics } from "@shared/schema";

export interface IStorage {
  // Product operations
  getProducts(): Promise<Product[]>;
  getProduct(id: number): Promise<Product | undefined>;
  getProductByCode(code: string): Promise<Product | undefined>;
  searchProducts(query: string): Promise<Product[]>;
  createProduct(product: InsertProduct): Promise<Product>;
  updateProduct(id: number, product: Partial<InsertProduct>): Promise<Product | undefined>;
  deleteProduct(id: number): Promise<boolean>;
  
  // Sell-in operations
  getSellIns(productId?: number, monthPartition?: string): Promise<SellIn[]>;
  createSellIn(sellIn: InsertSellIn): Promise<SellIn>;
  
  // Sell-through operations
  getSellThroughs(productId?: number, monthPartition?: string): Promise<SellThrough[]>;
  createSellThrough(sellThrough: InsertSellThrough): Promise<SellThrough>;
  
  // Analytics operations
  getProductAnalytics(productId?: number, monthPartition?: string): Promise<ProductAnalytics[]>;
  getOverallAnalytics(monthPartition?: string): Promise<OverallAnalytics>;
}

export class MemStorage implements IStorage {
  private products: Map<number, Product>;
  private sellIns: Map<number, SellIn>;
  private sellThroughs: Map<number, SellThrough>;
  private currentProductId: number;
  private currentSellInId: number;
  private currentSellThroughId: number;

  constructor() {
    this.products = new Map();
    this.sellIns = new Map();
    this.sellThroughs = new Map();
    this.currentProductId = 1;
    this.currentSellInId = 1;
    this.currentSellThroughId = 1;
  }

  async getProducts(): Promise<Product[]> {
    return Array.from(this.products.values());
  }

  async getProduct(id: number): Promise<Product | undefined> {
    return this.products.get(id);
  }

  async getProductByCode(code: string): Promise<Product | undefined> {
    return Array.from(this.products.values()).find(
      (product) => product.product_code === code
    );
  }

  async searchProducts(query: string): Promise<Product[]> {
    const lowercaseQuery = query.toLowerCase();
    return Array.from(this.products.values()).filter(
      (product) =>
        product.product_name.toLowerCase().includes(lowercaseQuery) ||
        product.product_code.toLowerCase().includes(lowercaseQuery) ||
        product.brand_name.toLowerCase().includes(lowercaseQuery) ||
        product.category_name.toLowerCase().includes(lowercaseQuery)
    );
  }

  async createProduct(insertProduct: InsertProduct): Promise<Product> {
    const id = this.currentProductId++;
    const now = new Date();
    const product: Product = {
      ...insertProduct,
      id,
      created_at: now,
      updated_at: now,
    };
    this.products.set(id, product);
    return product;
  }

  async updateProduct(id: number, updateData: Partial<InsertProduct>): Promise<Product | undefined> {
    const existingProduct = this.products.get(id);
    if (!existingProduct) {
      return undefined;
    }

    const updatedProduct: Product = {
      ...existingProduct,
      ...updateData,
      updated_at: new Date(),
    };
    this.products.set(id, updatedProduct);
    return updatedProduct;
  }

  async deleteProduct(id: number): Promise<boolean> {
    return this.products.delete(id);
  }

  async getSellIns(productId?: number, monthPartition?: string): Promise<SellIn[]> {
    let sellIns = Array.from(this.sellIns.values());
    
    if (productId) {
      sellIns = sellIns.filter(si => si.product_id === productId);
    }
    
    if (monthPartition) {
      sellIns = sellIns.filter(si => si.month_partition === monthPartition);
    }
    
    return sellIns;
  }

  async createSellIn(insertSellIn: InsertSellIn): Promise<SellIn> {
    const id = this.currentSellInId++;
    const sellIn: SellIn = {
      ...insertSellIn,
      id,
      created_at: new Date(),
    };
    this.sellIns.set(id, sellIn);
    return sellIn;
  }

  async getSellThroughs(productId?: number, monthPartition?: string): Promise<SellThrough[]> {
    let sellThroughs = Array.from(this.sellThroughs.values());
    
    if (productId) {
      sellThroughs = sellThroughs.filter(st => st.product_id === productId);
    }
    
    if (monthPartition) {
      sellThroughs = sellThroughs.filter(st => st.month_partition === monthPartition);
    }
    
    return sellThroughs;
  }

  async createSellThrough(insertSellThrough: InsertSellThrough): Promise<SellThrough> {
    const id = this.currentSellThroughId++;
    const sellThrough: SellThrough = {
      ...insertSellThrough,
      id,
      created_at: new Date(),
    };
    this.sellThroughs.set(id, sellThrough);
    return sellThrough;
  }

  async getProductAnalytics(productId?: number, monthPartition?: string): Promise<ProductAnalytics[]> {
    const products = productId 
      ? [this.products.get(productId)].filter(Boolean) as Product[]
      : Array.from(this.products.values());

    const analytics: ProductAnalytics[] = [];

    for (const product of products) {
      const sellIns = await this.getSellIns(product.id, monthPartition);
      const sellThroughs = await this.getSellThroughs(product.id, monthPartition);

      const sellInQuantity = sellIns.reduce((sum, si) => sum + si.quantity, 0);
      const sellThroughQuantity = sellThroughs.reduce((sum, st) => sum + st.quantity, 0);
      const totalRevenue = sellThroughs.reduce((sum, st) => sum + parseFloat(st.total_revenue || '0'), 0);
      const currentStock = sellInQuantity - sellThroughQuantity;
      const turnoverRate = sellInQuantity > 0 ? (sellThroughQuantity / sellInQuantity) * 100 : 0;

      analytics.push({
        product_id: product.id,
        product_name: product.product_name,
        product_code: product.product_code,
        brand_name: product.brand_name,
        sell_in_quantity: sellInQuantity,
        sell_through_quantity: sellThroughQuantity,
        turnover_rate: Math.round(turnoverRate * 10) / 10,
        total_revenue: totalRevenue,
        current_stock: currentStock,
      });
    }

    return analytics.sort((a, b) => b.total_revenue - a.total_revenue);
  }

  async getOverallAnalytics(monthPartition?: string): Promise<OverallAnalytics> {
    const sellIns = await this.getSellIns(undefined, monthPartition);
    const sellThroughs = await this.getSellThroughs(undefined, monthPartition);

    const totalSellIn = sellIns.reduce((sum, si) => sum + si.quantity, 0);
    const totalSellThrough = sellThroughs.reduce((sum, st) => sum + st.quantity, 0);
    const totalRevenue = sellThroughs.reduce((sum, st) => sum + parseFloat(st.total_revenue || '0'), 0);
    const averageTurnoverRate = totalSellIn > 0 ? (totalSellThrough / totalSellIn) * 100 : 0;

    return {
      total_sell_in: totalSellIn,
      total_sell_through: totalSellThrough,
      average_turnover_rate: Math.round(averageTurnoverRate * 10) / 10,
      total_revenue: totalRevenue,
    };
  }
}

export const storage = new MemStorage();

File: python_server/storage.py
from typing import List, Optional
from sqlalchemy import select
from .database import AsyncSessionLocal
from .db_models import ProductModel, SellInModel, SellThroughModel
from .models import (
    Product,
    InsertProduct,
    UpdateProduct,
    SellIn,
    InsertSellIn,
    SellThrough,
    InsertSellThrough,
    ProductAnalytics,
    OverallAnalytics,
)
import logging 

logger = logging.getLogger('uvicorn.error')


def to_schema(obj, schema_cls):
    if hasattr(schema_cls, "model_validate"):
        logger.debug(f"Model {obj} or {schema_cls}")
        return schema_cls.model_validate(obj, from_attributes=True)
    return schema_cls.from_orm(obj)


class SQLStorage:
    # Product operations
    async def get_products(self) -> List[Product]:
        async with AsyncSessionLocal() as session:
            result = await session.execute(select(ProductModel))
            return [to_schema(row, Product) for row in result.scalars().all()]

    async def get_product(self, pid: int) -> Optional[Product]:
        async with AsyncSessionLocal() as session:
            result = await session.get(ProductModel, pid)
            return to_schema(result, Product) if result else None

    async def get_product_by_code(self, code: str) -> Optional[Product]:
        async with AsyncSessionLocal() as session:
            stmt = select(ProductModel).where(ProductModel.product_code == code)
            result = await session.execute(stmt)
            row = result.scalar_one_or_none()
            return to_schema(row, Product) if row else None

    async def search_products(self, query: str) -> List[Product]:
        q = f"%{query.lower()}%"
        async with AsyncSessionLocal() as session:
            stmt = select(ProductModel).where(
                (ProductModel.product_name.ilike(q))
                | (ProductModel.product_code.ilike(q))
                | (ProductModel.brand_name.ilike(q))
                | (ProductModel.category_name.ilike(q))
            )
            result = await session.execute(stmt)
            return [to_schema(p, Product) for p in result.scalars().all()]

    async def create_product(self, data: InsertProduct) -> Product:
        async with AsyncSessionLocal() as session:
            obj = ProductModel(**data.dict())
            session.add(obj)
            await session.commit()
            await session.refresh(obj)
            return to_schema(obj, Product)

    async def update_product(self, pid: int, data: dict) -> Optional[Product]:
        async with AsyncSessionLocal() as session:
            obj = await session.get(ProductModel, pid)
            if not obj:
                return None
            for k, v in data.items():
                setattr(obj, k, v)
            await session.commit()
            await session.refresh(obj)
            return to_schema(obj, Product)

    async def delete_product(self, pid: int) -> bool:
        async with AsyncSessionLocal() as session:
            obj = await session.get(ProductModel, pid)
            if not obj:
                return False
            await session.delete(obj)
            await session.commit()
            return True

    # SellIn operations
    async def get_sell_ins(
        self, product_id: Optional[str] = None, month: Optional[str] = None
    ) -> List[SellIn]:
        async with AsyncSessionLocal() as session:
            stmt = select(SellInModel)
            if product_id is not None:
                stmt = stmt.where(SellInModel.product_id == product_id)
            if month:
                stmt = stmt.where(SellInModel.month_partition == month)
            result = await session.execute(stmt)
            return [to_schema(s, SellIn) for s in result.scalars().all()]

    async def create_sell_in(self, data: InsertSellIn) -> SellIn:
        async with AsyncSessionLocal() as session:
            obj = SellInModel(**data.dict())
            session.add(obj)
            await session.commit()
            await session.refresh(obj)
            return to_schema(obj, SellIn)

    # SellThrough operations
    async def get_sell_throughs(
        self, product_id: Optional[str] = None, month: Optional[str] = None
    ) -> List[SellThrough]:
        async with AsyncSessionLocal() as session:
            stmt = select(SellThroughModel)
            if product_id is not None:
                stmt = stmt.where(SellThroughModel.product_id == product_id)
            if month:
                stmt = stmt.where(SellThroughModel.month_partition == month)
            result = await session.execute(stmt)
            return [to_schema(s, SellThrough) for s in result.scalars().all()]

    async def create_sell_through(self, data: InsertSellThrough) -> SellThrough:
        async with AsyncSessionLocal() as session:
            obj = SellThroughModel(**data.dict())
            session.add(obj)
            await session.commit()
            await session.refresh(obj)
            return to_schema(obj, SellThrough)

    # Analytics operations
    async def get_product_analytics(
        self, product_id: Optional[str] = None, month: Optional[str] = None
    ) -> List[ProductAnalytics]:
        async with AsyncSessionLocal() as session:
            prod_stmt = select(ProductModel)
            if product_id is not None:
                prod_stmt = prod_stmt.where(ProductModel.id == product_id)
            products = (await session.execute(prod_stmt)).scalars().all()
            analytics: List[ProductAnalytics] = []
            for p in products:
                si_stmt = select(SellInModel).where(SellInModel.product_id == p.id)
                st_stmt = select(SellThroughModel).where(
                    SellThroughModel.product_id == p.id
                )
                if month:
                    si_stmt = si_stmt.where(SellInModel.month_partition == month)
                    st_stmt = st_stmt.where(SellThroughModel.month_partition == month)
                sell_ins = (await session.execute(si_stmt)).scalars().all()
                sell_throughs = (await session.execute(st_stmt)).scalars().all()
                sell_in_qty = sum(si.quantity for si in sell_ins)
                sell_through_qty = sum(st.quantity for st in sell_throughs)
                total_revenue = sum(float(st.total_revenue) for st in sell_throughs)
                current_stock = sell_in_qty - sell_through_qty
                turnover_rate = (sell_through_qty / sell_in_qty * 100) if sell_in_qty else 0
                analytics.append(
                    ProductAnalytics(
                        product_id=p.id,
                        product_name=p.product_name,
                        product_code=p.product_code,
                        brand_name=p.brand_name,
                        sell_in_quantity=sell_in_qty,
                        sell_through_quantity=sell_through_qty,
                        turnover_rate=round(turnover_rate, 1),
                        total_revenue=total_revenue,
                        current_stock=current_stock,
                    )
                )
            analytics.sort(key=lambda a: a.total_revenue, reverse=True)
            return analytics

    async def get_overall_analytics(self, month: Optional[str] = None) -> OverallAnalytics:
        async with AsyncSessionLocal() as session:
            si_stmt = select(SellInModel)
            st_stmt = select(SellThroughModel)
            if month:
                si_stmt = si_stmt.where(SellInModel.month_partition == month)
                st_stmt = st_stmt.where(SellThroughModel.month_partition == month)
            sell_ins = (await session.execute(si_stmt)).scalars().all()
            sell_throughs = (await session.execute(st_stmt)).scalars().all()
            total_sell_in = sum(si.quantity for si in sell_ins)
            total_sell_through = sum(st.quantity for st in sell_throughs)
            total_revenue = sum(float(st.total_revenue) for st in sell_throughs)
            avg_turnover = (total_sell_through / total_sell_in * 100) if total_sell_in else 0
            return OverallAnalytics(
                total_sell_in=total_sell_in,
                total_sell_through=total_sell_through,
                average_turnover_rate=round(avg_turnover, 1),
                total_revenue=total_revenue,
            )


storage = SQLStorage()
